shutdown/signal/signal.go

```diff
@@ -0,0 +1,45 @@
+package signal
+
+import (
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+)
+
+type Handler struct {
+	signalChannel chan os.Signal
+	wg            sync.WaitGroup
+}
+
+type Listener interface {
+	ShutdownSignalReceived()
+}
+
+func NewDefaultSignalHandler(listener Listener) *Handler {
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, syscall.SIGINT, syscall.SIGTERM)
+
+	return NewSignalHandler(signals, listener)
+}
+
+func NewSignalHandler(signalChannel chan os.Signal, listener Listener) *Handler {
+	handler := Handler{signalChannel: signalChannel}
+
+	go func() {
+		handler.waitForSignalAndCallListener(signalChannel, listener)
+	}()
+
+	return &handler
+}
+
+func (h *Handler) waitForSignalAndCallListener(signalChannel chan os.Signal, listener Listener) {
+	h.wg.Add(1)
+	_, _ = <-signalChannel
+	listener.ShutdownSignalReceived()
+	h.wg.Done()
+}
+
+func (h *Handler) WaitForSignal() {
+	h.wg.Wait()
```

Comment:
initally there was a time.Sleep(20ms) here (I assume is was added on purpose) why was it removed?

Reply:
i did not understand as to why this was needed